qiskit/opflow/gradients/gradient.py

```diff
@@ -41,20 +43,23 @@ class Gradient(GradientBase):
     # pylint: disable=signature-differs
     def convert(self,
                 operator: OperatorBase,
-                params: Union[ParameterVector, ParameterExpression, List[ParameterExpression]]
+                params: Optional[
+                    Union[ParameterVector, ParameterExpression, List[ParameterExpression]]] = None
                 ) -> OperatorBase:
         r"""
         Args:
             operator: The operator we are taking the gradient of.
-            params: params: The parameters we are taking the gradient with respect to.
+            params: The parameters we are taking the gradient with respect to. If not
+                explicitly passed, they are inferred from the operator and sorted by name.
 
         Returns:
             An operator whose evaluation yields the Gradient.
 
         Raises:
             ValueError: If ``params`` contains a parameter not present in ``operator``.
         """
-
+        if params is None:
+            params = sorted(operator.parameters, key=functools.cmp_to_key(_compare_parameters))
```

Comment:
@Cryoris Does this match the sorting that is done by QuantumCircuit. If not maybe it might be preferable since that new sorting order seems better

Reply:
yes, this is the same sorting:
https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/circuit/quantumcircuit.py#L1897

Reply:
Thanks for confirming, thats great!

I have to wonder though, whether, like QuantumCircuit, it would be better to return them sorted from the source, i.e. do this in opflow operators. Maybe @Cryoris has some thoughts/insight in that regard.

Reply:
Yes, ideally they should be sorted at source. This was one of the reasons why we chose sorting by name, so that we can extend this order to arbitrary opflow objects. 

I'm not sure how much work it is to apply this sorting globally in opflow. @molar-volume if you want you could try and do global sorting (if we're lucky it's as simple as sorting and returning a `ParameterView`, like in the circuit). But I'd be fine with doing the sorting inplace here and do the "sorting at source" in a follow-up.

Reply:
thanks, I’ll do the global sorting in this PR

Reply:
I am not sure how it fits the bigger picture, but changes to `ParameterView` in this PR have certain advantages:

1. parameters are sorted automatically in `ParameterView` (by overriding `__iter__`, which is used in `enumerate` , `for-in` loops and conversions).

2. `ParameterView` is subclass of `set`, so it is not breaking change and is consistent with parameters in quantum_info (and unit tests for operators were also expecting `set` as parameter return type)

3. Converting ordered parameters to list is as simple as https://github.com/molar-volume/qiskit-terra/blob/issue_5614/qiskit/opflow/gradients/gradient.py#L63

4. parameters of type `list` might have limitations for `OperatorBase` classes (if operators in `ListOp` share same parameters, `ListOp.parameters` would contain these parameters multiple times)

Reply:
If you agree with this approach, we might rename `ParameterView` to `ParameterSet` or `OrderedParameters`, but otherwise I think it is good to go.